growth-program/src/validation.rs

```diff
@@ -1,31 +1,40 @@
-use crate::{project::ProjectId, rewards::week_timekeeping::Week};
+use crate::{project::ProjectId, rewards::week_timekeeping::Week, GROWTH_SIGNATURE_PREFIX};
 
 multiversx_sc::imports!();
 
 pub type Signature<M> = ManagedByteArray<M, ED25519_SIGNATURE_BYTE_LEN>;
 pub const ED25519_SIGNATURE_BYTE_LEN: usize = 64;
 
+pub struct SignatureData<'a, M: ManagedTypeApi> {
+    pub caller: &'a ManagedAddress<M>,
+    pub project_id: ProjectId,
+    pub week: Week,
+    pub note: &'a ManagedBuffer<M>,
+}
+
 #[multiversx_sc::module]
-pub trait ValidationModule {
-    #[only_owner]
+pub trait ValidationModule: crate::project::ProjectsModule {
     #[endpoint(changeSigner)]
-    fn change_signer(&self, new_signer: ManagedAddress) {
-        self.signer().set(&new_signer);
+    fn change_signer(&self, project_id: ProjectId, new_signer: ManagedAddress) {
+        let caller = self.blockchain().get_caller();
+        self.require_sc_owner_or_project_owner(&caller, project_id);
+
+        self.signer(project_id).set(new_signer);
     }
 
     fn verify_signature(
         &self,
-        caller: &ManagedAddress,
-        project_id: ProjectId,
-        week: Week,
+        signature_data: SignatureData<Self::Api>,
         signature: &Signature<Self::Api>,
     ) {
         let mut data = ManagedBuffer::new();
-        let _ = caller.dep_encode(&mut data);
-        let _ = project_id.dep_encode(&mut data);
-        let _ = week.dep_encode(&mut data);
+        let _ = GROWTH_SIGNATURE_PREFIX.dep_encode(&mut data);
+        let _ = signature_data.project_id.dep_encode(&mut data);
+        let _ = signature_data.week.dep_encode(&mut data);
+        let _ = signature_data.caller.dep_encode(&mut data);
+        let _ = signature_data.note.dep_encode(&mut data);
```

Comment:
shouldn't we throw an error or something if encoding fails?

or is this handled by `verify_ed25519`?

Reply:
Encoding into a managed buffer can't fail. If it fails, there's something very wrong with the whole chain.

Reply:
Also, if encoding fails, the signature will be invalid anyway.